pkg/controller/oam/applicationconfiguration/render.go

```diff
@@ -106,13 +119,88 @@ func (r *components) Render(ctx context.Context, ac *v1alpha2.ApplicationConfigu
 			t.SetNamespace(ac.GetNamespace())
 
 			traits[i] = *t
-		}
 
-		workloads[i] = Workload{ComponentName: acc.ComponentName, Workload: w, Traits: traits}
+			traitDefinitionName := getCRDName(t)
+			var traitDef v1alpha2.TraitDefinition
+			err = r.client.Get(ctx, types.NamespacedName{
+				Name: traitDefinitionName,
+			}, &traitDef)
+			if err != nil {
+				return nil, errors.Wrapf(err, errFmtGetTraitDefinition, traitDefinitionName)
+			}
+			traitDefs = append(traitDefs, traitDef)
+		}
+		if err := SetWorkloadInstanceName(traitDefs, w, c); err != nil {
+			return nil, err
+		}
+		workloads[i] = Workload{ComponentName: acc.ComponentName, RevisionName: acc.RevisionName, Workload: w, Traits: traits}
 	}
 	return workloads, nil
 }
 
+// SetWorkloadInstanceName will set metadata.name for workload CR according to createRevision flag in traitDefinition
+func SetWorkloadInstanceName(traitDefs []v1alpha2.TraitDefinition, w *unstructured.Unstructured, c *v1alpha2.Component) error {
+	//Don't override the specified name
+	if w.GetName() != "" {
+		return nil
+	}
+	pv := fieldpath.Pave(w.Object)
+	if IsCreateRevision(traitDefs) {
+		// if trait aware revision, use revisionName as workload name
+		if err := pv.SetString(instanceNamePath, c.Status.LatestRevision); err != nil {
+			return errors.Wrapf(err, errSetValueForField, instanceNamePath, c.Status.LatestRevision)
+		}
+		return nil
+	}
+	// use component name as workload name, which means we will always use one workload for different revisions
+	if err := pv.SetString(instanceNamePath, c.GetName()); err != nil {
+		return errors.Wrapf(err, errSetValueForField, instanceNamePath, c.GetName())
+	}
+	w.Object = pv.UnstructuredContent()
+	return nil
+}
+
+// IsCreateRevision will check if any trait has createRevision flag, the appconfig should create a new workload instance
+func IsCreateRevision(traitDefs []v1alpha2.TraitDefinition) bool {
+	for _, td := range traitDefs {
+		if td.Spec.CreateRevision {
+			return true
+		}
+	}
+	return false
+}
+
+func getCRDName(u *unstructured.Unstructured) string {
+	apiVersion := u.GetAPIVersion()
+	kind := u.GetKind()
+	apis := strings.Split(apiVersion, "/")
+	if len(apis) < 2 {
+		return ""
+	}
+	// TODO(wonderflow) How could we find get specific crd name from CR object?
```

Comment:
This utility function is already implemented in the local-k8s repro, maybe we can move that here?

Reply:
thanks, I will use it